src/esp/physics/bullet/BulletRigidObject.cpp

```diff
@@ -36,34 +36,83 @@ bool BulletRigidObject::initializeScene(
 
-  bSceneArray_ = std::make_unique<btTriangleIndexVertexArray>();
-  bObjectShape_ = std::make_unique<btCompoundShape>();
-  constructBulletCompoundFromMeshes(Magnum::Matrix4{}, meshGroup, metaData.root,
-                                    false);
-  bObjectShape_->setMargin(0.0);
-  bSceneCollisionObjects_.emplace_back(std::make_unique<btCollisionObject>());
-  bSceneCollisionObjects_.back()->setCollisionShape(bObjectShape_.get());
-  bSceneCollisionObjects_.back()->setFriction(
-      physicsSceneAttributes.getDouble("frictionCoefficient"));
-  bSceneCollisionObjects_.back()->setRestitution(
-      physicsSceneAttributes.getDouble("restitutionCoefficient"));
-  bWorld->addCollisionObject(bSceneCollisionObjects_.back().get());
+  constructBulletSceneFromMeshes(Magnum::Matrix4{}, meshGroup, metaData.root);
+  for (auto& object : bSceneCollisionObjects_) {
+    object->setFriction(
+        physicsSceneAttributes.getDouble("frictionCoefficient"));
+    object->setRestitution(
+        physicsSceneAttributes.getDouble("restitutionCoefficient"));
+    bWorld->addCollisionObject(object.get());
+  }
 
   bWorld_ = bWorld;
-  syncPose();
   return true;
 }  // end BulletRigidObject::initializeScene
 
+void BulletRigidObject::constructBulletSceneFromMeshes(
+    const Magnum::Matrix4& transformFromParentToWorld,
+    const std::vector<assets::CollisionMeshData>& meshGroup,
+    const assets::MeshTransformNode& node) {
+  Magnum::Matrix4 transformFromLocalToWorld =
+      transformFromParentToWorld * node.transformFromLocalToParent;
+  if (node.meshIDLocal != ID_UNDEFINED) {
+    const assets::CollisionMeshData& mesh = meshGroup[node.meshIDLocal];
+
+    // SCENE: create a concave static mesh
+    btIndexedMesh bulletMesh;
+
+    Corrade::Containers::ArrayView<Magnum::Vector3> v_data = mesh.positions;
+    Corrade::Containers::ArrayView<Magnum::UnsignedInt> ui_data = mesh.indices;
+
+    //! Configure Bullet Mesh
+    //! This part is very likely to cause segfault, if done incorrectly
+    bulletMesh.m_numTriangles = ui_data.size() / 3;
+    bulletMesh.m_triangleIndexBase =
+        reinterpret_cast<const unsigned char*>(ui_data.data());
+    bulletMesh.m_triangleIndexStride = 3 * sizeof(Magnum::UnsignedInt);
+    bulletMesh.m_numVertices = v_data.size();
+    bulletMesh.m_vertexBase =
+        reinterpret_cast<const unsigned char*>(v_data.data());
+    bulletMesh.m_vertexStride = sizeof(Magnum::Vector3);
+    bulletMesh.m_indexType = PHY_INTEGER;
+    bulletMesh.m_vertexType = PHY_FLOAT;
+    bSceneArrays_.push_back(std::make_unique<btTriangleIndexVertexArray>());
+    bSceneArrays_.back()->addIndexedMesh(bulletMesh,
+                                         PHY_INTEGER);  // exact shape
+
+    //! Embed 3D mesh into bullet shape
+    //! btBvhTriangleMeshShape is the most generic/slow choice
+    //! which allows concavity if the object is static
+    bSceneShapes_.emplace_back(std::make_unique<btBvhTriangleMeshShape>(
```

Comment:
all of the `.back/.get` calls in this PR make it harder to read. How about setting up the unique_ptr first, and then at the very end adding it using `bSceneShapes_.emplace_back(std::move(ptr))`. This pattern (set up first, add to vector last) is also good to follow because if there is an error/exception during setup, then the invalid object is never added to the vector.

Reply:
Good suggestion 👍